ark/jsonschema/object.ts

```diff
@@ -0,0 +1,275 @@
+		ctx.reject({
+			expected: "a valid object JSON Schema",
+			actual:
+				"an object JSON Schema with 'required' array but no 'properties' object"
+		})
+	}
+
+	return {
+		optionalKeys: optionalKeys.map(key => ({
+			key,
+			value: parseJsonSchema(jsonSchema.properties![key]).internal
+		})),
+		requiredKeys: requiredKeys.map(key => ({
+			key,
+			value: parseJsonSchema(jsonSchema.properties![key]).internal
+		}))
+	}
+}
+
+const parseAdditionalProperties = (jsonSchema: JsonSchema.Object) => {
+	if (!("additionalProperties" in jsonSchema)) return
+
+	const properties =
+		jsonSchema.properties ? Object.keys(jsonSchema.properties) : []
+	const patternProperties = Object.keys(jsonSchema.patternProperties ?? {})
+
+	const additionalPropertiesSchema = jsonSchema.additionalProperties
+	if (additionalPropertiesSchema === true) return true
+	if (additionalPropertiesSchema === false) return false
+
+	const schemaDefinedKeys = rootSchema(
+		[...properties]
+			.map(key => ({ unit: key }))
+			.concat(
+				[...patternProperties].map(key => ({
+					domain: "string",
+					pattern: key
+				})) as never
+			)
+	)
+
+	const jsonSchemaObjectAdditionalPropertiesValidator = (
+		data: object,
+		ctx: Traversal
+	) => {
+		Object.keys(data).forEach(key => {
+			if (schemaDefinedKeys.allows(key))
+				// Not an additional property, so don't validate here
+				return
+
+			const additionalPropertyValidator = parseJsonSchema(
+				additionalPropertiesSchema
+			)
+
+			const value = data[key as keyof typeof data]
+			if (!additionalPropertyValidator.allows(value)) {
+				ctx.reject({
+					path: [key],
+					expected: `${additionalPropertyValidator.description}, since ${key} is an additional property.`,
+					actual: printable(value)
+				})
+			}
+		})
+		return !ctx.hasError()
+	}
+	return jsonSchemaObjectAdditionalPropertiesValidator
+}
+
+export const parseObjectJsonSchema: Type<
+	(In: JsonSchema.Object) => Out<Type<object, any>>,
+	any
+> = JsonSchemaScope.ObjectSchema.pipe((jsonSchema, ctx): Type<object> => {
+	const arktypeObjectSchema: Intersection.Schema<object> = {
+		domain: "object"
+	}
+
+	const { requiredKeys, optionalKeys } = parseRequiredAndOptionalKeys(
+		jsonSchema,
+		ctx
+	)
+	const patternPropertiesIndexes: Index.Schema[] =
+		parsePatternProperties(jsonSchema) ?? []
+
+	const parsedPropertyNamesSchema = parsePropertyNames(jsonSchema)
+	if (parsedPropertyNamesSchema === undefined) {
+		arktypeObjectSchema.required = requiredKeys
+		arktypeObjectSchema.optional = optionalKeys
+		arktypeObjectSchema.index = patternPropertiesIndexes
+	} else {
+		const propertyNamesIndex = {
+			signature: parsedPropertyNamesSchema,
+			value: type.unknown.internal
+		}
+
+		// Ensure all 'patternProperties' adhere to the 'propertyNames' schema
+		patternPropertiesIndexes.forEach(patternPropertyIndex => {
+			const patternPropertyAsIndexNode = node("index", patternPropertyIndex)
+			const propertyNamesAsIndexNode = node("index", propertyNamesIndex)
```

Comment:
We should instantiate this outside the loop

Reply:
Yep; done in 02df8aca